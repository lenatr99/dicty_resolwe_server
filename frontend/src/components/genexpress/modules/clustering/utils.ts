import { groupBy, reverse, round, sum } from 'lodash';
import { agnes, indices, Cluster } from './cluster';
import { ClusteringLinkageFunction, DistanceMeasure } from 'components/genexpress/common/constants';
import { ClusterNode, GeneExpression } from 'redux/models/internal';

export const euclidean = (v1: number[], v2: number[]) => {
    let powSum = 0;
    for (let index = 0; index < v1.length; index++) {
        const difference = v1[index] - v2[index];
        powSum += difference * difference;
    }
    return Math.sqrt(powSum);
};

export const pearson = (v1: number[], v2: number[]) => {
    const sum1 = sum(v1);
    const sum2 = sum(v2);
    const dat1 = v1.map((dataPoint) => Math.pow(dataPoint, 2));
    const dat2 = v2.map((dataPoint) => Math.pow(dataPoint, 2));
    const sum1sq = sum(dat1);
    const sum2sq = sum(dat2);

    const pData = v1.map((dataPoint, index) => dataPoint * v2[index]);

    const pSum = sum(pData);

    const num = pSum - (sum1 * sum2) / v1.length;
    const den = Math.sqrt(
        (sum1sq - Math.pow(sum1, 2) / v1.length) * (sum2sq - Math.pow(sum2, 2) / v1.length),
    );

    if (den === 0) {
        return 0;
    }

    return round(1.0 - num / den, 5);
};

const getRankArray = (arr: number[]) => {
    const sortedArr = [...arr].sort((a, b) => a - b);
    const rankMap = new Map<number, number>();

    for (let i = 0; i < sortedArr.length; i++) {
        const num = sortedArr[i];
        if (!rankMap.has(num)) {
            rankMap.set(num, i + 1);
        }
    }

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return arr.map((num) => rankMap.get(num)!);
};

const spearman = (arr1: number[], arr2: number[]) => {
    const n = arr1.length;

    const sortedIndexArr1 = getRankArray(arr1);
    const sortedIndexArr2 = getRankArray(arr2);

    let sumOfSquaredRankDifferences = 0;
    for (let i = 0; i < n; i++) {
        const rankDiff = sortedIndexArr1[i] - sortedIndexArr2[i];
        sumOfSquaredRankDifferences += rankDiff * rankDiff;
    }

    const spearmanDistance = (6 * sumOfSquaredRankDifferences) / (n * (n * n - 1));

    return spearmanDistance;
};

const distanceMeasureFunctionMap = {
    [DistanceMeasure.euclidean]: euclidean,
    [DistanceMeasure.pearson]: pearson,
    [DistanceMeasure.spearman]: spearman,
};

export const clusterByGenes = (
    sourceGeneExpressions: GeneExpression[],
    distanceMeasure: DistanceMeasure,
    linkageFunction: ClusteringLinkageFunction,
) => {
    const dataGroupedByGeneId = groupBy(sourceGeneExpressions, 'geneId');

    const genesWithoutValues: string[] = [];
    const genesWithAllConstantValues: string[] = [];
    const genesWithValues: string[] = [];
    const geneExpressions: number[][] = [];

    for (let index = 0; index < Object.keys(dataGroupedByGeneId).length; index++) {
        const geneId = Object.keys(dataGroupedByGeneId)[index];
        const items = dataGroupedByGeneId[geneId];

        if (items.some((item) => !isFinite(item.value)) || items.length === 0) {
            genesWithoutValues.push(geneId);
        } else if (items.every((item) => item.value === items[0].value)) {
            genesWithAllConstantValues.push(geneId);
        } else {
            genesWithValues.push(geneId);
            geneExpressions.push(items.map((item) => item.value));
        }
    }

    if (geneExpressions.length === 0) {
        return {
            withoutValues: genesWithoutValues,
            allConstantValues: genesWithAllConstantValues,
        };
    }

    const tree = agnes(geneExpressions, {
        method: linkageFunction,
        distanceFunction: distanceMeasureFunctionMap[distanceMeasure],
    });
    const intermediateTable: ClusterNode[] = [];

    // First element has order=0, that's why order value needs to be inverted
    // so that first element is shown at the top.
    const leafNodesOrder = reverse(indices(tree));
    // Get max distance from all nodes.
    const max = tree.height * 100;

    // Each clusterNode has to have a nodeIndex in order to track which gene is highlighted.
    let nonLeafNodeIndex = leafNodesOrder.length;

    const processNode = (node: Cluster, parent?: ClusterNode) => {
        const geneId = Object.keys(dataGroupedByGeneId)[node.index];
        if (node.isLeaf) {
            const yCoordinate = leafNodesOrder.indexOf(node.index);
            intermediateTable.push({
                nodeIndex: node.index,
                geneName: dataGroupedByGeneId[geneId][0].geneName,
                geneId: dataGroupedByGeneId[geneId][0].geneId,
                expressions: dataGroupedByGeneId[geneId],
                x: max,
                y: yCoordinate,
                parent,
            });
            return yCoordinate;
        } else if (node.children.length > 0) {
            const parentNode: ClusterNode = {
                x: max - node.height * 100,
                y: 0,
                nodeIndex: nonLeafNodeIndex++,
                parent,
            };
            intermediateTable.push(parentNode);

            let middle = 0;
            node.children.forEach((child) => {
                middle += processNode(child, parentNode);
            });

            middle = middle / 2;

            parentNode.y = middle;

            return middle;
        }

        return 0;
    };

    processNode(tree);

    const clusterNodes = intermediateTable.filter((node) => node.parent != null);
    const order = [
        ...leafNodesOrder.map((index) => genesWithValues[index]),
        ...genesWithoutValues,
        ...genesWithAllConstantValues,
    ];

    return {
        clusterNodes,
        order,
        withoutValues: genesWithoutValues,
        allConstantValues: genesWithAllConstantValues,
    };
};
