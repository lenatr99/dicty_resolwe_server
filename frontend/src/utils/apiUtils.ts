/**
 * Return deserialized response (type T) or null if response is not ok.
 * @param response Deserialized response json (as T).
 */
export const deserializeResponse = async <T>(response: Response): Promise<T> => {
    const responseJson = await response.json();
    return (responseJson.results != null ? responseJson.results : responseJson) as T;
};
