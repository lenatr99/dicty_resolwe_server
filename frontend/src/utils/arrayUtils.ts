import _ from 'lodash';

/**
 * Joins an array into "value1, value2, and value3"
 */
export const advancedJoin = (
    values: (string | undefined)[],
    delimiter = ', ',
    lastDelimiter = ', and ',
): string => {
    const nonNullValues = _.compact(values);
    if (_.size(nonNullValues) === 0) {
        return '';
    }
    if (_.size(nonNullValues) === 1) {
        return nonNullValues[0];
    }
    return `${_.initial(nonNullValues).join(delimiter)}${lastDelimiter}${
        _.last(nonNullValues) as string
    }`;
};
