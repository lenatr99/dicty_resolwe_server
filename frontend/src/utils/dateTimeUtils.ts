const dateFormat = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;

export const dateTimeReviver = (key: string, value: unknown): unknown => {
    if (typeof value === 'string' && dateFormat.test(value)) {
        return new Date(value);
    }

    return value;
};
