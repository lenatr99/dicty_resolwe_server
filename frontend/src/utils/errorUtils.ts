import { sentryCapture } from './sentryUtils';
import { ResponseError } from 'redux/models/internal';
import { addErrorSnackbar } from 'redux/stores/notifications';

export const handleError = (
    message: string,
    associatedObject: Record<string, unknown> | Error | ResponseError = {},
    suppressSentryCapture = false,
): ReturnType<typeof addErrorSnackbar> => {
    if (!suppressSentryCapture) {
        sentryCapture(message, associatedObject, 'error');
    }
    return addErrorSnackbar(message);
};

export const logError = (
    message: string,
    associatedObject: Record<string, unknown> | Error = {},
): void => {
    sentryCapture(message, associatedObject, 'error');
};

export const logInfo = (message: string, associatedObject: Record<string, unknown>): void => {
    sentryCapture(message, associatedObject, 'info');
};

export const logWarning = (message: string, associatedObject: Record<string, unknown>): void => {
    sentryCapture(message, associatedObject, 'warning');
};
