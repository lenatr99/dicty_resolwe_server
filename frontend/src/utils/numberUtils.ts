import _ from 'lodash';

/**
 * Generates an array of random numbers. If "randomNumberGenerator" function isn't given,
 * Math.random() is used.
 * @param arrayLength - How many numbers are in generated array.
 * @param randomNumberGenerator - Number generator function.
 */
export const generateRandomNumbers = (
    arrayLength: number,
    randomNumberGenerator: () => number = (): number => Math.random(),
): number[] => _.times(arrayLength, () => randomNumberGenerator());
