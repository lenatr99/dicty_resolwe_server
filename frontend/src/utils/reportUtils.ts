import _, { isArray } from 'lodash';
import * as vega from 'vega';

export const objectsArrayToTsv = (data: Record<string, unknown>[]): string => {
    if (_.isEmpty(data) || _.isEmpty(data[0])) {
        return '';
    }

    const headers = Object.keys(data[0]);
    return [headers, ...data]
        .map((item) => {
            if (isArray(item)) {
                return item.join('\t');
            }

            return headers.map((header) => item[header]).join('\t');
        })
        .join('\n');
};

export const vegaToPng = async (chartView: vega.View): Promise<string> => {
    const pngBase64 = (await chartView.toCanvas()).toDataURL('image/png');

    return pngBase64.substring(pngBase64.indexOf(',') + 1);
};

export const vegaToSvg = async (chartView: vega.View): Promise<Blob> => {
    return new Blob([await chartView.toSVG()], { type: 'image/svg+xml' });
};
