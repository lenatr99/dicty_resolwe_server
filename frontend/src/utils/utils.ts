/**
 * Asserts typeguard "never" to ensure exhaustive checks - e.g. switch covers all
 * types. It will trigger a compile error when a new check needs to be added to
 * cover a widened union type.
 *
 * Examples:
 * ```ts
 *   let value: 'a' | 'b';
 *
 *   if (value === 'a') {
 *   } else if (value === 'b') {
 *   } else {
 *       assertExhaustive(value);
 *   }
 *
 *   if (value === 'a') return 1;
 *   if (value === 'b') return 2;
 *   assertExhaustive(value);
 *
 *   switch (value) {
 *       case 'a': break;
 *       case 'b': break;
 *       default: assertExhaustive({value});
 *   }
 * ```
 * These examples will show a type error if type changes to 'a' | 'b' | 'c' and
 * case 'c' remains unhandled.
 */

import { logError } from './errorUtils';

export const assertExhaustive = (unexpectedValue: { [descriptiveKey: string]: never }): void => {
    logError(`Value ${JSON.stringify(unexpectedValue)} not handled despite type-exhaustive check`);
};
