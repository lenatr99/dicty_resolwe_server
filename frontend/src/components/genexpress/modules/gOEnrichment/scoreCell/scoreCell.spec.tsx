import React from 'react';
import { fireEvent, screen } from '@testing-library/react';
import GOEnrichmentScoreCell from './scoreCell';
import { customRender } from 'tests/test-utils';
import { generateGOEnrichmentRow } from 'tests/mock';
import { GOEnrichmentRow } from 'redux/models/internal';
import { formatNumber } from 'utils/math';

describe('gOEnrichmentScoreCell', () => {
    let goEnrichmentRow: GOEnrichmentRow;
    let asFragment: () => DocumentFragment;
    const value = 12;

    beforeEach(() => {
        goEnrichmentRow = generateGOEnrichmentRow(1);

        ({ asFragment } = customRender(
            <GOEnrichmentScoreCell value={value} data={goEnrichmentRow} />,
        ));
    });

    it('should render as default snapshot', () => {
        expect(asFragment()).toMatchSnapshot();
    });

    it('should call onMatchedGenesClick when user clicks on the link', () => {
        fireEvent.click(screen.getByText(formatNumber(value, 'short')));
    });
});
