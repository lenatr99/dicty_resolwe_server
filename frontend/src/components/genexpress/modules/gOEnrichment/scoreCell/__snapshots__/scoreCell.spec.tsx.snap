// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gOEnrichmentScoreCell > should render as default snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-jCbFiK hkFjrU"
  >
    <span
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="MuiLinearProgress-root-inRFPM jaInd MuiLinearProgress-root MuiLinearProgress-colorSecondary MuiLinearProgress-determinate sc-jzjKHc fYaqHH"
      role="progressbar"
    >
      <span
        class="MuiLinearProgress-bar1-ehbycK cgFaLI MuiLinearProgress-bar MuiLinearProgress-barColorSecondary MuiLinearProgress-bar1Determinate"
        style="transform: translateX(-100%);"
      />
    </span>
    <span
      class="sc-cBYhjr dIaTId"
    >
      12.00
    </span>
  </div>
</DocumentFragment>
`;
